packages/x-charts/src/internals/store/useCharts.ts

```diff
@@ -0,0 +1,111 @@
+import * as React from 'react';
+import { ChartStore } from '../plugins/utils/ChartStore';
+import {
+  ChartAnyPluginSignature,
+  ChartInstance,
+  ChartPlugin,
+  ChartPublicAPI,
+  ChartState,
+  ConvertSignaturesIntoPlugins,
+} from '../plugins/models';
+import { CHART_CORE_PLUGINS, ChartCorePluginSignatures } from '../plugins/corePlugins';
+import { UseChartBaseProps } from './useCharts.types';
+import { UseChartInteractionState } from '../plugins/featurePlugins/useChartInteraction/useChartInteraction.types';
+
+export function useChartApiInitialization<T>(
+  inputApiRef: React.MutableRefObject<T | undefined> | undefined,
+): T {
+  const fallbackPublicApiRef = React.useRef({}) as React.MutableRefObject<T>;
+
+  if (inputApiRef) {
+    if (inputApiRef.current == null) {
+      // eslint-disable-next-line react-compiler/react-compiler
+      inputApiRef.current = {} as T;
+    }
+    return inputApiRef.current;
+  }
+
+  return fallbackPublicApiRef.current;
+}
+
+let globalId = 0;
+
+export function useCharts<
+  TSignatures extends readonly ChartAnyPluginSignature[],
+  TProps extends Partial<UseChartBaseProps<TSignatures>>,
+>(inPlugins: ConvertSignaturesIntoPlugins<TSignatures>, props: TProps) {
+  type TSignaturesWithCorePluginSignatures = readonly [
+    ...ChartCorePluginSignatures,
+    ...TSignatures,
+  ];
+
+  const plugins = React.useMemo(
+    () =>
+      [
+        ...CHART_CORE_PLUGINS,
+        ...inPlugins,
+      ] as unknown as ConvertSignaturesIntoPlugins<TSignaturesWithCorePluginSignatures>,
+    [inPlugins],
+  );
+
+  const pluginParams = {}; // To generate when plugins use params.
+  const instanceRef = React.useRef({} as ChartInstance<TSignatures>);
+  const instance = instanceRef.current as ChartInstance<TSignatures>;
+  const publicAPI = useChartApiInitialization<ChartPublicAPI<TSignatures>>(props.apiRef);
+  const innerSvgRef: React.RefObject<SVGSVGElement> = React.useRef(null);
+
+  const storeRef = React.useRef<ChartStore<TSignaturesWithCorePluginSignatures> | null>(null);
+  if (storeRef.current == null) {
+    // eslint-disable-next-line react-compiler/react-compiler
+    globalId += 1;
+
+    const initialState = {
+      // TODO remove when the interaction moves to plugin
+      interaction: {
+        item: null,
+        axis: { x: null, y: null },
+      },
+      cacheKey: { id: globalId },
+    } as ChartState<TSignaturesWithCorePluginSignatures> & UseChartInteractionState;
+
+    plugins.forEach((plugin) => {
+      if (plugin.getInitialState) {
+        Object.assign(initialState, plugin.getInitialState({}));
+      }
+    });
+    storeRef.current = new ChartStore(initialState);
```

Comment:
@JCQuintas 

In this line the state is directly set because we are in the condition `storeRef.current == null` so we directly compute the initial state.

The discussion we had is about storing data from props into the store.

Since it's a ref we will need to use a `useEffect` to update the store. For example, the cartesian pluggin could contain to following code to make the xAxis available to selectors

```jsx
useEffect(() => {
  store.update(updateXAxis(xAxis))
}, [xAxis])
```

Which would cause 2 renders. One when user update their prop. A second one due to the store update.

So we trade potential double render with selectors

Reply:
Small note from the meeting: Unless the chart never uses any selector that contain this data AND the child component (the series? the dot? I don't know what consumes that :laughing: ) doesn't have props update or context update when this data changes from the parent.

In which case, we have a first render on the chart and then a second render on the child, but both only render once (except for the mount).